File: docs/examples/verify_webhook.py
"""
Trunk webhook signature verification helper (Python).

Usage:
    from verify_webhook import verify_trunk_webhook

    @app.post("/trunk-webhook")
    def handle_webhook(request):
        is_valid = verify_trunk_webhook(
            signature=request.headers.get("X-Trunk-Signature"),
            raw_body=request.data,
            webhook_secret=YOUR_WEBHOOK_SECRET,
        )
        if not is_valid:
            return {"error": "Invalid signature"}, 401
        # Process message...
"""

import hashlib
import hmac


def verify_trunk_webhook(
    signature: str | None,
    raw_body: bytes | str,
    webhook_secret: str,
) -> bool:
    if not signature or not signature.startswith("sha256="):
        return False

    expected = signature[7:]

    if isinstance(raw_body, str):
        raw_body = raw_body.encode("utf-8")

    computed = hmac.new(
        webhook_secret.encode("utf-8"),
        raw_body,
        hashlib.sha256,
    ).hexdigest()

    return hmac.compare_digest(expected, computed)

File: src/lib/verify-webhook.ts
/**
 * Webhook signature verification helper.
 * Use this to verify that incoming webhook payloads are from the Trunk relay.
 *
 * @example
 * ```ts
 * import { verifyTrunkWebhook } from "@usetrunk/verify-webhook";
 *
 * app.post("/trunk-webhook", (req, res) => {
 *   const isValid = await verifyTrunkWebhook(
 *     req.headers["x-trunk-signature"],
 *     req.body, // raw string
 *     YOUR_WEBHOOK_SECRET
 *   );
 *   if (!isValid) return res.status(401).send("Invalid signature");
 *   // Process message...
 * });
 * ```
 */

export async function verifyTrunkWebhook(
  signature: string | undefined | null,
  rawBody: string,
  webhookSecret: string
): Promise<boolean> {
  if (!signature || !signature.startsWith("sha256=")) return false;

  const expected = signature.slice(7);
  const computed = (await signTrunkWebhook(rawBody, webhookSecret)).slice(7);

  // Constant-time comparison
  if (expected.length !== computed.length) return false;
  let mismatch = 0;
  for (let i = 0; i < expected.length; i++) {
    mismatch |= expected.charCodeAt(i) ^ computed.charCodeAt(i);
  }
  return mismatch === 0;
}

export async function signTrunkWebhook(rawBody: string, webhookSecret: string): Promise<string> {
  const encoder = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(webhookSecret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );

  const sig = await crypto.subtle.sign("HMAC", key, encoder.encode(rawBody));
  const computed = Array.from(new Uint8Array(sig))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
  return `sha256=${computed}`;
}
